### PR TITLE
Add ability to specify custom epoch size

### PR DESCRIPTION
**Pull request by maksym-del**
Should fix https://github.com/allenai/allennlp/issues/4012

### CODE REVIEW

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -242,6 +266,10 @@ class GradientDescentTrainer(Trainer):
         precision training. Must be a choice of `"O0"`, `"O1"`, `"O2"`, or `"O3"`.
         See the Apex [documentation](https://nvidia.github.io/apex/amp.html#opt-levels-and-properties) for
         more details. If `None`, Amp is not used. Defaults to `None`.
+    batches_per_epoch: `int`, optional, (default = `None`)
```
Comment (matt-gardner) [inline allennlp/training/trainer.py:269]: We are trying to minimize the parameters that need to passed directly to the trainer.  I'm pretty sure you don't even need this, as you can just specify a `DataLoader`.  Can you make your data loader class inherit from our `DataLoader`, and register it as something?  In this PR, I'd like to see only modifications to data code, with no changes necessary to the trainer at all.

File: allennlp/training/trainer.py [hunk 2]
```diff
@@ -140,6 +140,30 @@ def __call__(
         raise NotImplementedError
 
 
+class CustomlySizedEpochDataLoader:
```
Comment (matt-gardner) [inline allennlp/training/trainer.py:143]: Can you move this inside `dataloader.py`?  Also, maybe name it `CustomEpochSizeDataLoader`?

File: allennlp/training/trainer.py (hunk 1 above)
Comment (maksym-del) [reply allennlp/training/trainer.py:269]: How do you like if I just add this optional parameter to existing Allennlp ```DataLoader```, so that default custom ```collate_fn``` is not the only reason we subclass it?

Comment (matt-gardner) [reply allennlp/training/trainer.py:269]: Sure, that seems reasonable enough to me.

Comment (maksym-del) [reply allennlp/training/trainer.py:269]: In this case, instead of specifying the new data loader with "type" keyword, just in order to be able to add a single parameter to the config file, the users will just get it as an additional field for free.

File: allennlp/data/dataloader.py [hunk 3]
```diff
@@ -36,6 +36,10 @@ class DataLoader(Registrable, data.DataLoader):
     # Instances which have _already_ been indexed.
     my_loader = DataLoader(dataset, batch_size=32, collate_fn=allennlp_collate)
     ```
+
+    Secondly, this class adds `batches_per_epoch` parameter which defines the epoch size in GPU batches
+    (as opposite to the effective batches obtained with Gradient Accumulation).
+    `None` value (default) means that the epoch size will equal the full pass through the data.
```
Comment (matt-gardner) [inline allennlp/data/dataloader.py:42]:
```suggestion
    Secondly, this class adds a `batches_per_epoch` parameter which, if given, determines the number
    of batches after which an epoch ends.  If this is `None`, then an epoch is set to be one full pass
    through your data.  You might use this if you have a very large dataset and want more frequent
    checkpoints and evaluations on validation data, for instance.
```

File: allennlp/tests/training/trainer_test.py [hunk 4]
```diff
@@ -139,6 +152,59 @@ def test_passing_trainer_multiple_gpus_raises_error(self):
                 self.model, self.optimizer, self.data_loader, num_epochs=2, cuda_device=[0, 1],
             )
 
+    def test_trainer_respects_epoch_size_equals_total(self):
+        num_epochs = 3
+        batches_per_epoch = len(self.data_loader_equal_epoch)
```
Comment (matt-gardner) [inline allennlp/tests/training/trainer_test.py:157]: All of these tests are identical and aren't really testing what you want, because there's nothing actually enforcing that the data loader did what you expected.  You're just reading a property here and are using this value in an assert below, in a way that should always be true of the trainer and is already tested.  I'd expect something like `assert batches_per_epoch == 3` here (or whatever the value should be).  Same for the tests below.

File: allennlp/training/trainer.py [hunk 5]
```diff
@@ -277,6 +277,7 @@ def __init__(
         self.model = model
 
         self.data_loader = data_loader
+
```
Comment (matt-gardner) [inline allennlp/training/trainer.py:280]: Can you just revert all changes to this file?
